homeassistant/components/switch/rainbird.py

```diff
@@ -0,0 +1,224 @@
+    password = config.get(CONF_PASSWORD)
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
+
+    from pyrainbird import RainbirdController
+
+    controller = RainbirdController(_LOGGER)
+    controller.setConfig(server, password)
+    _LOGGER.info("Rain Bird Controller set to " + str(server))
+
+    rbdevice = RainbirdDevice(hass, controller)
+    initialstatus = rbdevice.update()
+    if initialstatus == -1:
+        _LOGGER.error("Error getting state. Possible configuration issues")
+        raise PlatformNotReady
+    else:
+        _LOGGER.info("Initialized Rain Bird Controller")
+
+    entities = []
+    entities.append(rbdevice)
+    component.add_entities(entities)
+
+    devices = []
+    for dev_id, switch in config.get(CONF_SWITCHES).items():
+        devices.append(RainBirdSwitch(rbdevice, switch, dev_id))
+    add_devices(devices)
+    return True
+
+
+class RainBirdSwitch(SwitchDevice):
+    """Representation of a Rain Bird switch."""
+
+    def __init__(self, rb, dev, dev_id):
+        """Initialize a Rain Bird Switch Device."""
+        self._rainbird = rb
+        self._devid = dev_id
+        self._zone = int(dev.get(CONF_ZONE))
+        self._name = dev.get(CONF_FRIENDLY_NAME, "Sprinker %s" % self._zone)
+        self._state = self.get_device_status()
+        self._duration = dev.get(CONF_TRIGGER_TIME)
+        self._attributes = {
+            "duration": self._duration,
+            "zone": self._zone
+        }
+
+    @property
+    def device_state_attributes(self):
+        """Return state attributes."""
+        return self._attributes
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return True
+
+    @property
+    def name(self):
+        """Get the name of the switch."""
+        return self._name
+
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        self._state = True
+        self._rainbird.start_irrigation(self._zone, self._duration)
+        self._rainbird.setstate(self._zone)
+
+    def turn_off(self, **kwargs):
+        """Turn the switch off."""
+        self._state = False
+        self._rainbird.stop_irrigation()
+        self._rainbird.setstate(0)
+
+    def get_device_status(self):
+        """Get the status of the switch from Rain Bird Controller."""
+        return self._rainbird.state == self._zone
+
+    def update(self):
+        """Update switch status."""
+        self._state = self.get_device_status()
+
+    @property
+    def is_on(self):
+        """Return true if switch is on."""
+        return self._state
+
+
+class RainbirdDevice(Entity):
+    """Rain Bird Device."""
+
+    _state = -1
+
+    def __init__(self, hass, controller):
+        """Initialize the device."""
+        self.hass = hass
+        self.controller = controller
+        self._name = "Rainbird_Controller"
+        self._stations = {}
+
+        # For automation purposes add 2 services
+        def start_irrigation_call(call):
```

Comment:
These should be moved inside the `turn_on` and `turn_off` methods of the entity above, making this class (which is not an entity btw) obsolete. You should not register services for this but instead have switch.turn_on and switch.turn_off do the work.